Content translation integration branch

rafpaf opened the pull request:
The goal of this PR is to make it possible for admins to localize content in statically embedded questions and dashboards by uploading a translation dictionary.

Here's a screenshot:

<img width="500" alt="image" src="https://github.com/user-attachments/assets/0aa845c4-be6d-4d3c-aa4c-13e9cdad02c2" />

## Try it out

🚗 Take it for a spin: You can try this out yourself in the PR env: https://metaboat.slack.com/archives/C089KHNCY3X/p1748631227007689

## Context

The code in this PR was reviewed, PR by PR, but now needs an additional thorough, holistic review. I'd also be grateful for input from the Querying, DashViz, Embedding teams, since this PR modifies their code.

Some additional BE unit tests are in a [downstream PR](https://github.com/metabase/metabase/pull/58747).

rafpaf committed:
Localize headings in statically embedded dashboards

rafpaf committed:
Content translation project integration branch

rafpaf commented:
* **#56557** <a href="https://app.graphite.dev/github/pr/metabase/metabase/56557?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#56880** <a href="https://app.graphite.dev/github/pr/metabase/metabase/56880?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#56879** <a href="https://app.graphite.dev/github/pr/metabase/metabase/56879?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/metabase/metabase/56879?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

rafpaf committed:
Add BE unit tests

rafpaf committed:
This introduces logic that validates content translation dictionaries
when they are uploaded.

A valid dictionary must have:

* Three fields per row
* A valid locale in the first column
* One translation per string per locale (for example, you can only provide one translation of "Category" into German - you can't have two rows that begin "de,Category,")
* Cells containing strings no longer than 255 characters

rafpaf committed:
Restore BE validation error logic

rafpaf committed:
Don't verify summaries

rafpaf committed:
Fix missing import

rafpaf committed:
Translate tab name if it is not editable

rafpaf committed:
Add e2e test for localizing dashboard tab names

rafpaf committed:
Localize content of dashboard text cards (#58497)

rafpaf committed:
Restore CSS

rafpaf committed:
Fix unit tests

rafpaf committed:
Fix unit tests

rafpaf committed:
Move some files around

rafpaf committed:
Content translation: Localize categorical field values in static-embedded dashboards and questions (#58504)

rafpaf committed:
Fix test

dpsutton committed:
Content translation cleanup (#58553)

* remove full response for simple map

```diff
-  {:status http-status-ok
-   :headers {"Content-Type" "application/json"}
-   :body (json/encode {:success true})})
+  {:success true})
```
this is equivalent and happens automatically

* just pass a file object

gonna change this as i don't love the api but it's better than the map
passing in the filename that wasn't needed. This is single argument

* pass rows in, not a file

* bit of cleanup in the dictionary creation namespace

rafpaf committed:
Content translation: localize question names and descriptions in static embedding (#57564)

dpsutton committed:
endpoint in embedding for translations

```
-- when no locale you get nothing and you will be happy
❯ http get 'localhost:3000/api/embed/dictionary/eyJhbGciOiJIUzI1NiJ9.eyJyZXNvdXJjZSI6eyJkYXNoYm9hcmQiOjExfSwicGFyYW1zIjp7fX0.7R1vvqgYnERw05m_22B5S-XLG-cZIZW3BOLY8XjgBf8'
HTTP/1.1 204 No Content

-- with french you get french
❯ http get 'localhost:3000/api/embed/dictionary/eyJhbGciOiJIUzI1NiJ9.eyJyZXNvdXJjZSI6eyJkYXNoYm9hcmQiOjExfSwicGFyYW1zIjp7fX0.7R1vvqgYnERw05m_22B5S-XLG-cZIZW3BOLY8XjgBf8?locale=fr'
HTTP/1.1 200 OK
...

[
    {
        "id": 393,
        "locale": "fr",
        "msgid": "Orders over time",
        "msgstr": "l'orders sur temps"
    },
    {
        "id": 394,
        "locale": "fr",
        "msgid": "Orders displayed over tim",
        "msgstr": "Orders qui nous displayons sur temps"
    },
    ...
    ]

```

rafpaf committed:
wip

dpsutton committed:
remove old unauthed api route in favor of embedding route

previous commit added embedding route, this removes the old routes

rafpaf committed:
Remove failing but low priority sorting unit test - ideally we should mock out translateContentString

rafpaf committed:
Content translation: validate uploads (#57544)

rafpaf committed:
Content translation: downloadable dictionary (#58607)

rafpaf committed:
Add setup functions for content dictionary jwt endpoints

rafpaf committed:
Support Jwt authentication for static embeds

Simplify CSV logic

rafpaf committed:
Fix lint stuff

rafpaf committed:
Fix import

rafpaf committed:
Remove unused import

rafpaf committed:
Update doc string

rafpaf committed:
lint

rafpaf committed:
lint

rafpaf committed:
Remove console.logs

rafpaf committed:
Shorten file name, remove unneeded check

rafpaf committed:
Fix some be tests

rafpaf committed:
Fix e2e and unit tests, and kondo issues. Limit scope of content translations to static embedding only.

rafpaf committed:
Reduce diff

rafpaf committed:
Gate the content translation endpoint setter

rafpaf committed:
Small tweaks

rafpaf committed:
Clarify doc

rafpaf committed:
Reduce diff

rafpaf committed:
Reduce diff

rafpaf committed:
Improve var name

rafpaf committed:
Remove note

rafpaf committed:
Increase version number

rafpaf committed:
Fix whitespace

rafpaf committed:
Fix types

rafpaf committed:
Fix some types

rafpaf committed:
Reduce diff

rafpaf committed:
Fix e2e tests

rafpaf committed:
Restore changes to TableInteractive

rafpaf committed:
Rollback changes to ListField unit tests

rafpaf committed:
simplify unit tests involving content translations

rafpaf committed:
Mark test as flaky: click-behavior.cy.spec: 'should allow settin dashboard and saved question as custom destination for different columns'

rafpaf committed:
Fix type issues

rafpaf committed:
try to fix flake

rafpaf committed:
I'm confused why these unit tests don't work

rafpaf committed:
wip: fixing unit tests

rafpaf committed:
Use ngettext

rafpaf committed:
Fix imports

rafpaf committed:
Fix types and tests

rafpaf committed:
Prevent console error when dashcard errors

rafpaf committed:
wip: move content translation embedding endpoint

rafpaf committed:
Remove comments

rafpaf committed:
wip

rafpaf committed:
Gate content translation utilities behind token feature

rafpaf committed:
Remove unneeded code

rafpaf committed:
Reduce jankiness in content translation form

rafpaf committed:
Remove unneeded comment

rafpaf commented on enterprise/frontend/src/metabase-enterprise/content_translation/utils.ts:
```diff
@@ -0,0 +1,157 @@
+      data: { ...singleSeries.data, rows: translatedRows },
+    };
+  });
+};
```
For performance it might be sensible to early-exit this function if there are no translations

rafpaf committed:
Reduce diff

rafpaf committed:
Reduce diff

rafpaf committed:
Early exit from content translation functions when there are no translations

rafpaf committed:
Improve copy in content translation form

rafpaf committed:
wip: improving BE unit tests

rafpaf committed:
Improve be unit tests

rafpaf committed:
wip: improving BE unit tests

rafpaf committed:
wip: adding unit tests for embedded ct dict endpoint, moving FE embedded endpoint setter to enterprise

rafpaf committed:
Rename func

rafpaf committed:
almost there with embedded dictionary route test

rafpaf committed:
Fix be unit test

rafpaf committed:
Remove todo

rafpaf committed:
Use ! and remove note

rafpaf committed:
Fix jest tests

rafpaf committed:
Fix migrations

rafpaf committed:
Remove log/infos

rafpaf committed:
wip: fixing be unit testse

rafpaf committed:
Fix some be unit tests

rafpaf committed:
Dictionary be tests pass

rafpaf committed:
Reduce diff

rafpaf committed:
Only re-sort SingleSelectListWidget options if there are content translatoins

rafpaf committed:
Remove change to View

rafpaf committed:
Fix kondo errors

rafpaf committed:
wip: memoizing translateFieldValuesInHoveredObject

rafpaf committed:
Don't use a hook

rafpaf committed:
Memoize translateHoveredObject by turning it into a hook

rafpaf committed:
wip: bringing back with-clean-translations

rafpaf committed:
bringing back with-clean-translations

rafpaf committed:
wip: trying to fix be unit tests

rafpaf committed:
Fix bug and doc

rafpaf committed:
fix be unit tests in routes_test

iethree commented on frontend/src/metabase/parameters/components/widgets/ParameterFieldWidget/FieldValuesWidget/ListField/utils.ts:
```diff
@@ -1,3 +1,31 @@
+ * Returns true when the option matches the filter. If the option has an
+ * available content translation, the filter is tested against this translation
+ **/
+export function optionMatchesFilter(
+  option: FieldValue | undefined,
+  filter: string,
+  translate?: ContentTranslationFunction,
+) {
+  // If filter has no diacritics (like äéĩøů), let's ignore diacritics in the
+  // option
+  const ignoreDiacritics = filter === removeDiacritics(filter);
+
+  return option?.some((value) => {
+    const stringifiedValue = String(value ?? "");
+    const maybeTranslatedValue =
+      translate?.(stringifiedValue) || stringifiedValue;
+    const normalizedValue = ignoreDiacritics
+      ? removeDiacritics(maybeTranslatedValue)
+      : maybeTranslatedValue;
+    return normalizedValue.toLowerCase().includes(filter);
+  });
 }
```
this is very fancy, but I'm wondering what performance implications it has on large field lists

iethree commented on frontend/src/metabase/parameters/components/widgets/ParameterFieldWidget/FieldValuesWidget/ListField/utils.unit.spec.ts:
❤️  nice tests

iethree commented on frontend/src/metabase/parameters/components/widgets/ParameterFieldWidget/Value/Value.tsx:
❤️  typescript

iethree commented on frontend/src/metabase/visualizations/visualizations/Heading/Heading.tsx:
```diff
@@ -46,6 +49,8 @@ export function Heading({
   const preventDragging = (e: MouseEvent<HTMLInputElement>) =>
     e.stopPropagation();
 
+  const translatedText = useMemo(() => tc(settings.text), [settings.text, tc]);
```
I think `tc` should self-memoize, right?

iethree commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -11545,6 +11545,8 @@ databaseChangeLog:
                     nullable: true
                     unique: true
 
+
+
```
🗑️

iethree commented on frontend/test/__support__/utils.ts:
```diff
@@ -53,3 +53,22 @@ export async function findRequests(method: "PUT" | "POST" | "DELETE" | "GET") {
   return Promise.all(reqs);
 }
+
+/**
+ * Asserts that a Jest expect() assertion fails. This function is useful for
+ * ensuring that the page is NOT asynchronously updated. It ensures that the
+ * callback function repeatedly throws an error until the timeout elapses
+ *
+ * NOTE: To verify in a unit test that something does not occur, it may be more
+ * efficient to spy on a function and verify that it is not called.
+ *
+ * @see https://stackoverflow.com/questions/68118941/how-to-wait-for-something-not-to-happen-in-testing-library
+ * */
+export const assertNeverPasses = async (
+  fn: () => void | Promise<void>,
+  expectedError: string | RegExp = /Timed out in waitFor/,
+  options: waitForOptions = {},
+) => {
+  options.timeout ??= 200;
+  await expect(waitFor(fn, options)).rejects.toThrow(expectedError);
+};
```
this isn't used in this PR. If you feel strongly about having it available, I'd open a separate PR for it

iethree reviewed (changes requested):
❤️  This is a significant piece of work, very well done. 

Many of my comments are quibbles/suggestions. The few I think really need to be addressed are:

- We should revisit the use of plugins to disable things within the enterprise folder
- A few files need some more test coverage
- You're hitting some _very_ hot paths with field values and parameters, I want more confidence (maybe via unit tests?) that we're not doing insane things to QB performance here

heypoom commented on enterprise/frontend/src/metabase-enterprise/content_translation/index.ts:
```diff
@@ -0,0 +1,28 @@
+    // This gets overwritten in static embedding
+    contentTranslationDictionaryUrl: null,
+    setEndpoints: (encodedToken: string) => {
+      PLUGIN_CONTENT_TRANSLATION.contentTranslationDictionaryUrl = `/api/ee/embedded-content-translation/dictionary/${encodedToken}`;
```
That's the JWT token for static embedding; it stores the static embedding config e.g. dashboard id, question id, and locked parameters.

sloansparger commented on enterprise/frontend/src/metabase-enterprise/content_translation/utils.ts:
```diff
@@ -0,0 +1,175 @@
+    return msgid;
+  }
+
+  const msgstr = dictionary?.find(
+    (row) => row.locale === locale && row.msgid === msgid,
+  )?.msgstr;
```
I don't think we can do an O(n * m) lookup per render (n = number of items to translate, m = size of your dictionary) for views with large datasets. Something like `type TCDictionary<locale, Record<msgid, translation>>` is going to serve us better I think O(n).

And as always with perf... measure! Ideally using as real world of a dictionary as possible. With a tiny enough dictionary I'd expect that this code as is would beat out my suggestion.

sloansparger commented on enterprise/frontend/src/metabase-enterprise/content_translation/use-translate-content.ts:
Nothing is standing out to me.If this hook is going to be used hundreds/thousands of times I kind of wonder about how fast the `useListContentTranslationsQuery` usage will be, but that's it.

rafpaf commented on enterprise/frontend/src/metabase-enterprise/api/content-translation.ts:
```diff
@@ -0,0 +1,52 @@
+      >({
+        query: (params) => ({
+          method: "GET",
+          // The URL is currently only set in static embedding
+          url: PLUGIN_CONTENT_TRANSLATION.contentTranslationDictionaryUrl,
```
At the moment I can only think of worse approaches

rafpaf commented on frontend/src/metabase/i18n/hooks.ts:
```diff
@@ -0,0 +1,27 @@
+export const useTranslateContent = <
+  T = string | null | undefined,
+>(): ContentTranslationFunction => {
+  return PLUGIN_CONTENT_TRANSLATION.useTranslateContent<T>();
```
Thinking about this one. The `useTranslateContent` in use-translate-content.ts returns a useCallback with two dependencies - the locale and the dictionary. If I memoize the function here, it won't re-run when the dictionary GET request returns results, so I think the function will be stale.

johnswanson commented on enterprise/backend/src/metabase_enterprise/content_translation/routes.clj:
```diff
@@ -0,0 +1,68 @@
+
+(api.macros/defendpoint :get "/csv"
+  "Provides content translation dictionary in CSV"
+  [_route-params
```
pretty sure this can just be `[]` (you don't need to list the unused args here)

johnswanson commented on enterprise/backend/test/metabase_enterprise/content_translation/dictionary_test.clj:
```diff
@@ -0,0 +1,173 @@
+(defn- get-translations
+  "Get all translations from the database."
+  []
+  (t2/select :model/ContentTranslation))
```
should this be ordered? alphabetized or something? just thinking that we might want to give these to people in a consistent order

johnswanson commented on enterprise/backend/src/metabase_enterprise/content_translation/routes.clj:
```diff
@@ -0,0 +1,68 @@
+
+;; The maximum size of a content translation dictionary is 1.5MiB
+;; This should equal the maxContentDictionarySizeInMiB variable in the frontend
+(def ^:private max-content-translation-dictionary-size (* 1.5 1024 1024))
```
I always like including units in var names when it's potentially relevant, ie `max-content-translation-dictionary-size-bytes`.

You could also do something like

```
(def ^:private max-content-translation-dictionary-size-mb 1.5)
(def ^:private max-content-translation-dictionary-size-bytes (* max-content-translation-dictionary-size-mb 1024 1024))
```

then later when you're throwing the too-big exception you don't need to worry about dividing by `(* 1024 1024)`. Not blocking though.

johnswanson commented on enterprise/backend/src/metabase_enterprise/content_translation/routes.clj:
```diff
@@ -0,0 +1,68 @@
+    (when-not (instance? java.io.File file)
+      (throw (ex-info (tru "No file provided") {:status-code 400})))
+    (with-open [rdr (io/reader file)]
+      (let [[_header & rows] (csv/read-csv rdr)]
```
what happens if the file doesn't look like a CSV? I guess that will just end up as errors from `import-translations!`?

johnswanson commented on enterprise/backend/src/metabase_enterprise/embed/content_translation/api.clj:
```diff
@@ -0,0 +1,37 @@
+(ns metabase-enterprise.embed.content-translation.api
```
thoughts about putting this in `metabase-enterprise.content-translation.api.embed` instead? Not sure we want an entirely new separate module for embedding, especially since it seems to wholly depend on content translation?

johnswanson commented on enterprise/backend/src/metabase_enterprise/embed/content_translation/api.clj:
```diff
@@ -0,0 +1,37 @@
+  [{:keys [token]} :- [:map
+                       [:token string?]]
+   {:keys [locale]}]
+  (log/info "hit embed ct endpoint")
```
I don't think this is necessary, pretty sure we already have logging on all requests.

johnswanson commented on enterprise/backend/src/metabase_enterprise/embed/content_translation/api.clj:
```diff
@@ -0,0 +1,37 @@
+    (throw (ex-info (str (tru "Locale is required.")) {:status-code 400}))))
+
+(defn- +require-content-translation [handler]
+  (fn [request respond raise]
```
Maybe we could make `(embedding.jwt/unsign token)` happen in a middleware like `+require-content-translation` and then apply it to all routes in this namespace? Juuuuust in case :)

johnswanson commented on enterprise/backend/test/metabase_enterprise/content_translation/dictionary_test.clj:
```diff
@@ -0,0 +1,173 @@
+(defn- count-translations
+  "Count the number of translations in the database."
+  []
+  (count (t2/select :model/ContentTranslation)))
```
`(t2/count :model/ContentTranslation)` is going to do a `SELECT count(*)` instead of counting the results of `SELECT *`, should be faster!

johnswanson reviewed (commented):
I'm not quite done looking through, but thought I'd leave what I have for now. I'll take another look tomorrow. So far it's looking good!

rafpaf commented on enterprise/frontend/src/metabase-enterprise/parameters/components/ParameterWidget/ParameterWidget.unit.spec.tsx:
```diff
@@ -0,0 +1,91 @@
+    it("(EE with token) should translate legend", async () => {
+      setup({
+        ...setupOptions,
+        hasEnterprisePlugins: true,
```
Good point. I've put the OSS/EE tests in different files to avoid any problems like this

rafpaf committed:
Address more FE and BE feedback

rafpaf committed:
Fix whitespace

rafpaf committed:
Don't restore data editing api

rafpaf committed:
Fix merge conflicts

rafpaf committed:
Remove embedded routes test

johnswanson committed:
Add a test for bad token on embedding

Also refactored things a bit to make things clearer

- Instead of having a dynamic var with the current key, just bind the
current key to a var so we can explicitly sign things with it (or not,
as the case may be)

- test that the error messages look right for the various error
cases (to be sure we're not e.g. accidentally hitting the "no locale"
case when we meant to hit the "no token" case)

rafpaf committed:
Fix util tests

rafpaf committed:
Add unit tests for other hooks that perform content translation

rafpaf committed:
Fix test name

rafpaf committed:
Add unit test for useSortByContentTranslation

rafpaf committed:
Fix lint

rafpaf committed:
Sort modules in config, remove duplicate, assign team to content-translation

rafpaf committed:
Fix CI, improve layout in ContentTranslationConfiguration

rafpaf committed:
Fix be unit tests

rafpaf committed:
Fix redundant let expression

rafpaf committed:
wip: trying to fix a flake

rafpaf committed:
Avoid socket error

rafpaf committed:
Normalize locales on upload

rafpaf committed:
Reject invalid CSV with useful message

iethree reviewed (approved):
🎉 Frontend looks good to me 🎉

rafpaf committed:
Remove interceptContentTranslationRoutes

rafpaf commented on enterprise/frontend/src/metabase-enterprise/api/content-translation.ts:
```diff
@@ -0,0 +1,51 @@
+
+          return {
+            method: "POST",
+            url: contentTranslationEndpoints.uploadDictionary,
```
`contentTranslationEndpoints.getDictionary` is the one that gets set in static embedding. It has a jsdoc comment on it

rafpaf committed:
Translate card names using a simpler method that also works for Scalar visualizations

rafpaf committed:
Update useLocale usage

rafpaf committed:
Remove unneeded comments

rafpaf committed:
Remove unneeded comments

rafpaf committed:
Reduce diff

johnswanson commented:
FYI the failing tests are due to a rename: you can just swap `mt/with-empty-h2-app-db` for `mt/with-empty-h2-app-db!`

rafpaf committed:
Fix imports

rafpaf committed:
Simplify hasTranslations, using function equality rather than a property

rafpaf committed:
Address feedback

rafpaf committed:
Remove logs

rafpaf committed:
Fix test

rafpaf committed:
wip

rafpaf committed:
Simplify

rafpaf committed:
Fix be unit tests

rafpaf committed:
Add clearer doc string

rafpaf committed:
Fix bug where invalid locales lead to duplication errors

rafpaf committed:
Fix tests

rafpaf committed:
Translate noncategorical field values, too

rafpaf committed:
Don't add query-reference-validation

rafpaf committed:
Fix be unit test

rafpaf committed:
Fix lint

rafpaf committed:
Fix imports

rafpaf committed:
Fix imports

rafpaf committed:
Fix SingleSelectListField widget

rafpaf committed:
Fix unit tests

rafpaf committed:
Fix e2e tests

rafpaf committed:
Mark tests as flaky

github-automation-metabase committed:
Update Loki Snapshots
